cmd/clusterctl/client/tree/discovery.go

```diff
@@ -105,35 +123,130 @@ func Discovery(ctx context.Context, c client.Client, namespace, name string, opt
+			resourceSetRefObject := ObjectReferenceObject(&corev1.ObjectReference{
+				Kind:       "ClusterResourceSet",
+				Namespace:  cluster.Namespace,
+				Name:       binding.ClusterResourceSetName,
+				APIVersion: addonsv1.GroupVersion.String(),
+			})
+			tree.Add(resourceSetGroup, resourceSetRefObject, ObjectMetaName("ClusterResourceSet"))
+		}
 	}
+}
 
-	workers := VirtualObject(cluster.Namespace, "WorkerGroup", "Workers")
-	tree.Add(cluster, workers)
+func addControlPlane(cluster *clusterv1.Cluster, controlPlane *unstructured.Unstructured, tree *ObjectTree, options DiscoverOptions) {
+	tree.Add(cluster, controlPlane, ObjectMetaName("ControlPlane"), GroupingObject(true))
+
+	if options.ShowTemplates {
+		// Add control plane infrastructure ref using spec fields guaranteed in contract
+		infrastructureRef, found, err := unstructured.NestedMap(controlPlane.UnstructuredContent(), "spec", "machineTemplate", "infrastructureRef")
+		if err == nil && found {
+			infrastructureObjectRef := &corev1.ObjectReference{
+				Kind:       infrastructureRef["kind"].(string),
+				Namespace:  infrastructureRef["namespace"].(string),
+				Name:       infrastructureRef["name"].(string),
+				APIVersion: infrastructureRef["apiVersion"].(string),
+			}
+
+			machineTemplateRefObject := ObjectReferenceObject(infrastructureObjectRef)
+			var templateParent client.Object
+			if options.AddTemplateVirtualNode {
+				templateParent = VirtualObject(cluster.Namespace, "TemplateGroup", "Templates", to.StringPtr(controlPlane.GetName()+"-templates"))
+				tree.Add(controlPlane, templateParent, ZOrder(1))
+			} else {
+				templateParent = controlPlane
+			}
+			tree.Add(templateParent, machineTemplateRefObject, ObjectMetaName("MachineInfrastructureTemplate"))
+		}
+	}
+}
 
+func addMachineDeploymentToObjectTree(ctx context.Context, c client.Client, cluster *clusterv1.Cluster, workers *unstructured.Unstructured, machinesList *clusterv1.MachineList, tree *ObjectTree, options DiscoverOptions, addMachineFunc func(parent client.Object, m *clusterv1.Machine)) error {
 	// Adds worker machines.
 	machinesDeploymentList, err := getMachineDeploymentsInCluster(ctx, c, cluster.Namespace, cluster.Name)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	machineSetList, err := getMachineSetsInCluster(ctx, c, cluster.Namespace, cluster.Name)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	for i := range machinesDeploymentList.Items {
 		md := &machinesDeploymentList.Items[i]
 		addOpts := make([]AddObjectOption, 0)
-		if !options.ShowMachineSets {
+		if !options.ShowMachineSets { // TODO: does this need to be a list?
 			addOpts = append(addOpts, GroupingObject(true))
 		}
 		tree.Add(workers, md, addOpts...)
 
+		if options.ShowTemplates {
+			var templateParent client.Object
+			if options.AddTemplateVirtualNode {
+				templateParent = VirtualObject(cluster.Namespace, "TemplateGroup", "Templates", to.StringPtr(md.GetName()+"-templates"))
```

Comment:
@ykakarap I fixed the issue you mentioned earlier. It's due to the fact that the tree uses UIDs as keys, and VirtualObjects have a deterministic UID based on the kind, namespace, and name which caused a collision when we have multiple machine deployments. 

I amended the function definition to include an optional UID parameter where we could have collisions. Alternatively, we could append a random string to the UID like how k8s does it, but our current testing framework for discovery.go relies on deterministic UIDs.